### DISCUSSION TITLE
dynamic current control for DSMR fed SmartEVSE's

### DISCUSSION THREAD

**Discussion by dingo35** [serkri/SmartEVSE-3#221] [Announcements]
In the Netherlands many users feed their SmartEVSE's MainsMeter data through the EM_API REST-API; a problem that arises is that the original feed of MainsMeter data happens every 1-2 seconds (either through modbus kWh meter or Sensorbox), but DSMR meters typically provide new data every 5-10 seconds.

Current behaviour of the current control loop is:
-down regulation: if MaxMains is exceed by amount X, the charging current is immediately reduced with X, until it falls below MinCurrent (typically 6A), in which case it pauses charging.
-up regulation: if MaxMains is not reached yet, unused capacity Y is calculated (MaxMains - current usage), and the charging current is increased by Y/4 .
This control loop is run every 1-2 seconds.

NEW behaviour in this version:
[phases_update_timer.zip](https://github.com/serkri/SmartEVSE-3/files/13632798/phases_update_timer.zip)

-down regulation remains the same
-up regulation: the control loop is run ONLY on update of the MainsMeter data.
-timeout of the communication is set from 10s to 11 seconds.

Please test! It is a debug version so it is possible to telnet to your device and see what's going on....

**Comment by Imaginous** [comment]
I have installed the firmware, have not tested the up regulation yet (have to install the hardware for real, it's still on my testbench).

One thing I noticed that the time is in 12 hour format instead of 24 hour format. 
It would be nice if it is selectable... or do I miss something ;)

**Comment by dingo35** [comment]
Do you mean the time on the LCD screen or the webserver?

At my location it is all in 24h notation, it is probably dependent on your local OS/browser/Timezone settings.

**Reply by Imaginous** [reply]
I mean in the webserver (webpage).
I'm also in the Netherlands. As I can remember the original firmware showed 24H.

If the weather permits I want to install everything tomorrow and will do some tests.

**Reply by Imaginous** [reply]
It seems a Windows "feature"... On my phone it still shows 24H.

Sorry for the interruption.
Will try rebooting Windows... No idea why it happened after the update.

**Reply by Imaginous** [reply]
It happens al my Windows systems in Edge and Chrome.
So it seems something has changed.

Al machines use 24H time notation.

**Reply by dingo35** [reply]
I'm all Linux here, but found access to a Windows 10 machine, that still gives 24h notation when surfing to my SmartEVSE webserver. Nonetheless I checked my code and for displaying the Time: field it uses "toLocaleTimeString"; so it explicitly asks the browser to use the locally defined parameters. Google "windows system locale", Microsoft might have it more complicated than only 12h/24h time notation...

Furthermore, lets keep this thread on topic: dynamic current control

**Comment by Imaginous** [comment]
Back on topic. Finished installing my SmartEVSE.

The device seems to step up nicely with the "beat" of the API input.

Have to do the 32 amps test tomorrow.

**Comment by Imaginous** [comment]
Maybe I don't understand something.

Mains 35A
Max 32A

When SMART is selected and the car is charging at 32A and the water kettle turns on I understood that the charging current is lowered in steps of 6A (till a minimum of 6A) so the total current is below the mains 35A.

But for some reason it completely stops (while charging at 16A would be possible) and starts counting down again and restarts charging ramping up slowly from 6A.

Why does it stop?

**Comment by Imaginous** [comment]
![Screenshot_20231216-141307_Chrome](https://github.com/serkri/SmartEVSE-3/assets/68418842/ed494c49-29f2-4e08-a345-23ea68ea10a8)

**Comment by Imaginous** [comment]
![Screenshot_20231216-141636_Chrome](https://github.com/serkri/SmartEVSE-3/assets/68418842/5f447f5d-2ef0-44db-8d93-af48faf5a381)

**Comment by dingo35** [comment]
You are right, in SMART mode the charge current is supposed to go down, not in 6A steps but immediately to the available current, e.g. if the kettle uses 8A and there are no other users in the house, chargecurrent should drop to 35-8=27A.

I suspect one of the safety-measures that were built in the original firmware, is kicking in, too early since it is not synchronized to the phases update .

Could you make a verbose telnet-log of the kick-down moment and post that here?

**Reply by Imaginous** [reply]
Do I have to use Putty to get the verbose telnet-log?
Haven't done that before.

I will give it a try tomorrow, have to head out now.

**Reply by dingo35** [reply]
Telnet used to be included in windows on the command line; if its not there, putty would be your way to telnet from a Windows machine.

**Comment by Imaginous** [comment]
I found the culprit, there was a glitch in my software bridge relaying the HomeWizzard P1 data to the SmartEVSE.

As soon as an over current situation occurred it would sent me a warning message... but a faulty IF statement also stopped sending data to SmartEVSE.

So SmartEVSE shuts down due to the fact of temporarily not receiving data, which is 101% wanted and expected behavior.

So this feature works fine.

Is it already in Version 1.8.0?
